### ISSUE TITLE
Incorrect Error with NamedTuple slices ? 

### ISSUE THREAD

**Issue by Carreau** [python/mypy#8776]
Assuming the following file:
```
$ cat foo.py
from typing import NamedTuple, List, Tuple, Sequence

from typing import NamedTuple, List, Sequence


class NT(NamedTuple):
    matched_text: str
    matches: Sequence[str]   # Sequence will not match List below
    matches_origin: int


def f() -> Tuple[str, List[str]]: # List will not match sequence above.
    nt = NT("a", ["a", "b"], 1)
    return nt[:2]
```

Got
```
$ mypy foo.py
foo.py:14: error: Incompatible return value type (got "NT", expected "Tuple[str, List[str]]")
Found 1 error in 1 file (checked 1 source file)
```

I would have expected `(got "Tuple[str, Seq[str]]", expected...)` not `got "NT"`)

Looks like the inference is good, if I replace Sequence by List or vice versa, It's all fine, just the error forget about the slicing.

**Comment by EkanshMangal** [comment]
Hello Sir, I want to start contributing in open source.
Can i fix this issue?

**Comment by graingert** [comment]
This error is worse than just invalid error reporting, mypy thinks a sliced NamedTuple is still an instance of the NamedTuple rather than a plain tuple:

eg it considers this valid:
```python
from typing import NamedTuple

class Example(NamedTuple):
    foo: int
    bar: int

def foo() -> Example:
    return Example(1, 2)[:2]
```

https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=29ce22a2fc602683eecc8322abcdd272

**Comment by graingert** [comment]
the error reporting is particularly confusing when expecting a NamedTuple return value:
mypy reports: `main.py:8: error: Incompatible return value type (got "Example", expected "Example")  [return-value]`
```python
from typing import NamedTuple

class Example(NamedTuple):
    foo: int
    bar: int

def foo() -> Example:
    return Example(1, 2)[:1]
```
https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=5c1e943d3c033ce955df109bb745d674

**Comment by graingert** [comment]
mypy also considerers this valid:

```python
class Foo(tuple[int, int]):
    pass


f: Foo = Foo((1, 1))[:]
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=7cf7292436cd2b3a09acaf4c7c3fdd33
